src/library/scala/collection/immutable/Vector.scala

```diff
@@ -120,12 +120,13 @@ sealed abstract class Vector[+A] private[immutable] (private[immutable] final va
 
   override def iterableFactory: SeqFactory[Vector] = Vector
 
-  override final def length: Int =
-    if(this.isInstanceOf[BigVector[_]]) this.asInstanceOf[BigVector[_]].length0
-    else prefix1.length
+  override final def length: Int = this match {
+    case value: BigVector[_] => value.length0
+    case _ => prefix1.length
+  }
```

Comment:
I didn't measure the impact but the pattern match creates some unnecessary bytecode instructions:

```
public final int length();
    descriptor: ()I
    Code:
       0: aload_0
       1: instanceof    #485                // class scala/collection/immutable/BigVector
       4: ifeq          15
       7: aload_0
       8: checkcast     #485                // class scala/collection/immutable/BigVector
      11: invokevirtual #488                // Method scala/collection/immutable/BigVector.length0:()I
      14: ireturn
      15: aload_0
      16: invokevirtual #490                // Method prefix1:()[Ljava/lang/Object;
      19: arraylength
      20: ireturn
```

```
public final int length();
    descriptor: ()I
    Code:
       0: aload_0
       1: instanceof    #490                // class scala/collection/immutable/BigVector
       4: ifeq          18
       7: aload_0
       8: checkcast     #490                // class scala/collection/immutable/BigVector
      11: invokevirtual #493                // Method scala/collection/immutable/BigVector.length0:()I
      14: istore_1
      15: goto          24
      18: aload_0
      19: invokevirtual #495                // Method prefix1:()[Ljava/lang/Object;
      22: arraylength
      23: istore_1
      24: iload_1
      25: ireturn
```

Reply:
Hehe.  Instead of just `ireturn`ing, it `istore_1`s, `goto`s, `iload_1`s and _then_ `ireturn`s.